src/qibo/derivative.py

```diff
@@ -1,7 +1,180 @@
+from qibo.models.error_mitigation import CDR, calibration_matrix
+from qibo.symbols import I, Symbol, Z
+
+
+class Parameter:
+    """Object which allows complex gate parameters. Several trainable parameter
+    and possibly features are linked through a lambda function which returns the
+    final gate parameter"""
+
+    def __init__(self, func, trainablep, featurep=None):
+        self._trainablep = trainablep
+        self._featurep = featurep
+        self.nparams = len(trainablep)
+        self.lambdaf = func
+
+    def _apply_func(self, fixed_params=None):
+        """Applies lambda function and returns final gate parameter"""
+        params = []
+        if self._featurep is not None:
+            params.append(self._featurep)
+        if fixed_params:
+            params.extend(fixed_params)
+        else:
+            params.extend(self._trainablep)
+        return self.lambdaf(*params)
+
+    def _update_params(self, trainablep=None, feature=None):
+        """Update gate trainable parameter and feature values"""
+        if trainablep:
+            self._trainablep = trainablep
+        if feature:
+            self._featurep = feature
+
+    def get_params(self, trainablep=None, feature=None):
+        """Update values with trainable parameter and calculate current gate parameter"""
+        self._update_params(trainablep=trainablep, feature=feature)
+        return self._apply_func()
+
+    def get_indices(self, start_index):
+        """Return list of respective indices of trainable parameters within
+        the optimizer's trainable parameter list"""
+        return [start_index + i for i in range(self.nparams)]
+
+    def get_fixed_part(self, trainablep_idx):
+        """Retrieve parameter constant unaffected by a specific trainable parameter"""
+        params = [0] * self.nparams
+        params[trainablep_idx] = self._trainablep[trainablep_idx]
+        return self._apply_func(fixed_params=params)
+
+    def get_scaling_factor(self, trainablep_idx):
+        """Get scaling factor multiplying a specific trainable parameter"""
+        params = [0] * self.nparams
+        params[trainablep_idx] = 1.0
+        return self._apply_func(fixed_params=params)
+
+
+def create_hamiltonian(qubit, nqubit, backend):
+    """
+    Creates appropriate Hamiltonian for a given list of qubits
+    Args:
+        qubit: qubit numbers whose states we are interested in
+        nqubit: total number of qubits, which determines size of Hamiltonian
+    Return:
+        hamiltonian: SymbolicHamiltonian
+    """
+    if not isinstance(qubit, list):
+        qubit = [qubit]
+
+    hams = []
+
+    for i in range(nqubit):
+        if i in qubit:
+            hams.append(Z(i))
+        else:
+            hams.append(I(i))
+
+    # create Hamiltonian
+    obs = np.prod([Z(i) for i in range(1)])
+    hamiltonian = SymbolicHamiltonian(obs, backend=backend)
+
+    return hamiltonian
+
+
+def error_mitigation(circuit, hamiltonian, backend, noise_model):
```

Comment:
The usage of error mitigation should be a user choice, unrelated to the calculation of a derivative. We can think to involve this as option into the optimization module, but I would suggest to remove this from the differentiation one.

Reply:
I moved this into the optimizer.py file